File: tests/pytests/utils/TestPylithVersion.py
# =================================================================================================
# This code is part of PyLith, developed through the Computational Infrastructure
# for Geodynamics (https://github.com/geodynamics/pylith).
#
# Copyright (c) 2010-2025, University of California, Davis and the PyLith Development Team.
# All rights reserved.
#
# See https://mit-license.org/ and LICENSE.md and for license information. 
# =================================================================================================

import unittest

from pylith.testing.TestCases import make_suite
from pylith.utils.utils import PylithVersion

class TestPylithVersion(unittest.TestCase):

    def test_isRelease(self):
        isRelease = PylithVersion.isRelease()


    def test_version(self):
        version = PylithVersion.version()
        # Check that version is of the form X.X.X
        import re
        match = re.search("[0-9]+\\.[0-9]+\\.[0-9]+", version)
        self.assertFalse(match is None)


    def test_gitVersion(self):
        revision = PylithVersion.gitRevision()
        if PylithVersion.isRelease():
            self.assertEqual("unknown", revision)
        else:
            # Check that revision is of the form v2.1.3-16-g9323114
            import re
            match = re.search("v[0-9]+\\.[0-9]+\\.[0-9]+", revision)
            if match is None:
                match = re.search("v[0-9]+\\.[0-9]+\\.[0-9]+-[0-9]+-g[0-9,a-z]+", revision)
            self.assertFalse(match is None)


    def test_gitHash(self):
        tag = PylithVersion.gitHash()
        if PylithVersion.isRelease():
            self.assertEqual("unknown", tag)
        else:
            # Check form of hash
            import re
            match = re.search("[0-9,a-z]+", tag)
            self.assertFalse(match is None)


    def test_gitDate(self):
        value = PylithVersion.gitDate()
        if PylithVersion.isRelease():
            self.assertEqual("unknown", value)
        else:
            # Check form of datetime
            import datetime
            fields = value.split()
            d = datetime.datetime.strptime(fields[0], "%Y-%m-%d")
            t = datetime.datetime.strptime(fields[1], "%H:%M:%S")


    def test_gitBranch(self):
        branch = PylithVersion.gitBranch()
        if PylithVersion.isRelease():
            self.assertEqual("unknown", branch)
        else:
            self.assertFalse(len(branch) == 0)


def load_tests(loader, tests, pattern):
    TEST_CLASSES = [TestPylithVersion]
    return make_suite(TEST_CLASSES, loader)


if __name__ == "__main__":
    unittest.main(verbosity=2)


# End of file 

File: tests/libtests/utils/TestPylithVersion.cc
// -*- C++ -*-
//
// ----------------------------------------------------------------------
//
// Brad T. Aagaard, U.S. Geological Survey
//
// This code was developed as part of the Computational Infrastructure
// for Geodynamics (http://geodynamics.org).
//
// Copyright (c) 2010-2022 University of California, Davis
//
// See LICENSE.md for license information.
//
// ----------------------------------------------------------------------
//

#include <portinfo>

#include "pylith/utils/PylithVersion.hh" // Test subject

#include <string> // USES std::string()
#include <string.h> // USES strlen()

#include "catch2/catch_test_macros.hpp"

// ------------------------------------------------------------------------------------------------
namespace pylith {
    namespace utils {
        class TestPylithVersion;
    }
}

class pylith::utils::TestPylithVersion {
    // PUBLIC METHODS /////////////////////////////////////////////////////////////////////////////
public:

    /// Test isRelease()
    static
    void testIsRelease(void);

    /// Test version()
    static
    void testVersion(void);

    /// Test gitRevision()
    static
    void testGitRevision(void);

    /// Test gitHash()
    static
    void testGitHash(void);

    /// Test gitDate()
    static
    void testGitDate(void);

    /// Test gitBranch()
    static
    void testGitBranch(void);

}; // class TestPylithVersion

// ------------------------------------------------------------------------------------------------
TEST_CASE("TestPylithVersion::testIsRelease", "[TestPylithVersion]") {
    pylith::utils::TestPylithVersion::testIsRelease();
}
TEST_CASE("TestPylithVersion::testVersion", "[TestPylithVersion]") {
    pylith::utils::TestPylithVersion::testVersion();
}
TEST_CASE("TestPylithVersion::testGitRevision", "[TestPylithVersion]") {
    pylith::utils::TestPylithVersion::testGitRevision();
}
TEST_CASE("TestPylithVersion::testGitHash", "[TestPylithVersion]") {
    pylith::utils::TestPylithVersion::testGitHash();
}
TEST_CASE("TestPylithVersion::testGitDate", "[TestPylithVersion]") {
    pylith::utils::TestPylithVersion::testGitDate();
}
TEST_CASE("TestPylithVersion::testGitBranch", "[TestPylithVersion]") {
    pylith::utils::TestPylithVersion::testGitBranch();
}

// ------------------------------------------------------------------------------------------------
// Test isRelease()
void
pylith::utils::TestPylithVersion::testIsRelease(void) { // testIsRelease
#if PYLITH_RELEASE_VERSION
    CHECK(PylithVersion::isRelease());
#else
    CHECK(!PylithVersion::isRelease());
#endif
} // testIsRelease


// ------------------------------------------------------------------------------------------------
// Test version()
void
pylith::utils::TestPylithVersion::testVersion(void) { // testVersion
    CHECK(std::string(PYLITH_VERSION) == std::string(PylithVersion::version()));
} // testVersion


// ------------------------------------------------------------------------------------------------
// Test gitRevision()
void
pylith::utils::TestPylithVersion::testGitRevision(void) { // testGitRevision
#if PYLITH_RELEASE_VERSION
    CHECK(std::string("unknown") == std::string(PylithVersion::gitRevision()));
#else
    // Git revision should be of the form vX.X.X-gXXXXX.
    const char* rev = PylithVersion::gitRevision();
    CHECK('v' == rev[0]);
#endif
} // testGitRevision


// ------------------------------------------------------------------------------------------------
// Test gitHash()
void
pylith::utils::TestPylithVersion::testGitHash(void) { // testGitHash
#if PYLITH_RELEASE_VERSION
    CHECK(std::string("unknown") == std::string(PylithVersion::gitHash()));
#else
    // Git hash should contain lower case and numbers.
    const char* hash = PylithVersion::gitHash();
    const int len = strlen(hash);
    for (int i = 0; i < len; ++i) {
        const int value = int(hash[i]);
        assert((value >= int('0') && value <= int('9')) || (value >= int('a') && value <= int('z')));
    } // for
#endif
} // testGitHash


// ------------------------------------------------------------------------------------------------
// Test gitDate()
void
pylith::utils::TestPylithVersion::testGitDate(void) { // testGitDate
#if PYLITH_RELEASE_VERSION
    CHECK(std::string("unknown") == std::string(PylithVersion::gitDate()));
#else
    const char* datetime = PylithVersion::gitDate();
    CHECK(strlen(datetime) > 0);
#endif
} // testGitDate


// ------------------------------------------------------------------------------------------------
// Test gitBranch()
void
pylith::utils::TestPylithVersion::testGitBranch(void) { // testGitBranch
#if PYLITH_RELEASE_VERSION
    CHECK(std::string("unknown") == std::string(PylithVersion::gitBranch()));
#else
    const char* branch = PylithVersion::gitBranch();
    CHECK(strlen(branch) > 0);
#endif
} // testGitBranch


// End of file
